### PR TITLE
Quality standards

### PR DESCRIPTION
**Pull request by vhirtham**
## Changes
_Describe changes in this PR_

## Related Issues
Closes # (add issue numbers)

## Checks
- [x] updated CHANGELOG.md
- [x] updated tests
- [x] updated doc/
- [x] update example/tutorial notebooks 
- [x] ~~review implementation of `QualityStandard` class in light of [the new documentation](https://asdf.readthedocs.io/en/latest/asdf/extending/resources.html#extending-resources) (Keyword: `DirectoryResourceMapping`)~~ will be refactored in 0.5.0

### CODE REVIEW

**Comment by vhirtham** [comment]
The tutorial text is still "work in progress", but the relevant code sections are there and are fully functional.

**Comment by vhirtham** [comment]
I didn't strip the tutorial from all outputs so that it can be inspected on GitHub. The documentation builds are currently failing for multiple reasons and they have to be fixed first. If this is done, I will remove the outputs from the notebook.

**Comment by CagtayFabry** [comment]
this PR should be up to date now with master/2.8

**Comment by marscher** [comment]
The tutorial notebook uses obsolete functions from weldx.asdf.utilities. Should I replace the file handling with WeldxFile patterns? @CagtayFabry

**Comment by CagtayFabry** [comment]
> The tutorial notebook uses obsolete functions from weldx.asdf.utilities. Should I replace the file handling with WeldxFile patterns? @CagtayFabry

if you have some headroom to look into it that sounds good 👍 
we should do an in depth review to finalize once @vhirtham is back :)

**Comment by marscher** [comment]
> 
> 
> > The tutorial notebook uses obsolete functions from weldx.asdf.utilities. Should I replace the file handling with WeldxFile patterns? @CagtayFabry
> 
> if you have some headroom to look into it that sounds good 👍
> we should do an in depth review to finalize once @vhirtham is back :)

I will put it on my list then :)

**Comment by vhirtham** [comment]
> > > The tutorial notebook uses obsolete functions from weldx.asdf.utilities. Should I replace the file handling with WeldxFile patterns? @CagtayFabry
> > 
> > 
> > if you have some headroom to look into it that sounds good 👍
> > we should do an in depth review to finalize once @vhirtham is back :)
> 
> I will put it on my list then :)

Should be fixed now :)

**Comment by vhirtham** [comment]
The basics are now fixed, but there are some things to consider. I summarized them in #361 and #362 . I will reread the tutorial and see if there is something missing. Will also check the coverage if there are some things that lack proper testing. Otherwise, we can merge for now and fix the things mentioned in the issues later.

File: tutorials/quality_standards.ipynb
Comment (marscher) [inline tutorials/quality_standards.ipynb:46]: this can be imported directly from weldx

File: tutorials/quality_standards.ipynb
Comment (marscher) [inline tutorials/quality_standards.ipynb:80]: which link should be fixed?

**Comment by marscher** [comment]
The tutorial still contains several todo's. I guess these should be fixed before merging or will there be a follow up PR?

File: tutorials/quality_standards.ipynb
Comment (vhirtham) [reply tutorials/quality_standards.ipynb:80]: Guess I was referring to the link to the standards. It's been a while since I wrote the tutorial. The problem is that the link doesn't work if you run the tutorial in a notebook, because the Kernel's working directory is (at least on my system) `weldx/tutorials`. In the documentation, the tutorials run in `weldx/doc/tutorials` because they are copied to this location.
So the relative paths to the corresponding `.rst` file differ. Ideas?

**Comment by vhirtham** [comment]
When updating the tutorial I came across some things that are still left to do. They don't have to be included in this PR, but we should decide what we want to include and which topics get a new issue as a reminder.

**Things left to discuss:**

- how to solve incompatible notebook/sphinx links (see discussion above)
- content of the manifest file, so that the corresponding tutorial section can be completed

**Things that are left to be done:**

- Write a guideline on how a quality standard should be structured
- tutorial about `wx_metadata`

**Comment by CagtayFabry** [comment]
> When updating the tutorial I came across some things that are still left to do. They don't have to be included in this PR, but we should decide what we want to include and which topics get a new issue as a reminder.
> 
> **Things left to discuss:**
> 
> * how to solve incompatible notebook/sphinx links (see discussion above)
> * content of the manifest file, so that the corresponding tutorial section can be completed
> 
> **Things that are left to be done:**
> 
> * Write a guideline on how a quality standard should be structured
> * tutorial about `wx_metadata`

my thoughts:
- the notebook link issue is a nuisance, maybe we find a solution. For now keep links fixed in web documentation
- I think the manifest file is fine as it suits our purpose, we can review later in light of:
https://asdf.readthedocs.io/en/al-503-document-2.8-features/asdf/extending/manifests.html#extension-manifests
- guideline: we should create s complete usage demo using the template:
https://github.com/BAMWelDX/quality-standard-template

**Review by CagtayFabry — CHANGES_REQUESTED** [review]
I think the code and implementation is fine, just the documentation should be updated:

- fill out missing todos (as concise as possible)
- change `GenericEquipment` for `MeasurementEquipment`?
- add link to the cookiecutter-template repo

**Comment by vhirtham** [comment]
> I think the code and implementation is fine, just the documentation should be updated:
> 
> * fill out missing todos (as concise as possible)
> * change `GenericEquipment` for `MeasurementEquipment`?
> * add link to the cookiecutter-template repo

Done. Just take a look at the [diff of the last commit](https://github.com/BAMWelDX/weldx/pull/211/commits/ee70440519d3bfc0a5c29352c068801b18d6651c) for fast review. The other commits where just renaming and fixing stuff after merging master into this branch.

**Comment by vhirtham** [comment]
Almost forgot: We need to make the cookie-cutter repo public. Otherwise the links wont work

File: tutorials/quality_standards.ipynb
Comment (marscher) [inline tutorials/quality_standards.ipynb:1]: The tutorials needs to be pep8/black formatted.

**Comment by marscher** [comment]
@vhirtham  Is it intended, that a QS directory only have to contain manifests with the same base name and only differing versions? I tried to add another manifest for testing purposes, e.g. test_standard2-1.0.0.yaml and then I get a ValueError("Inconsistent naming of manifest files") raised in config.py:56.

**Comment by vhirtham** [comment]
> @vhirtham Is it intended, that a QS directory only have to contain manifests with the same base name and only differing versions? I tried to add another manifest for testing purposes, e.g. test_standard2-1.0.0.yaml and then I get a ValueError("Inconsistent naming of manifest files") raised in config.py:56.

Yes, it is. It's also mentioned somewhere in the tutorial ;)

**Comment by marscher** [comment]
There is one mistake in the tag name of the test-standard. It should be "measurement_equipment" now, not "generic_equipment".

The problem we are seeing is due to the fact, that the load_schema function in asdf is lru_cached, so a second call to with the same url (e.g. enable a standard resolves to a tag url) is resolved to the cache. The local config context circumvents this problem, but I have not yet figured, if it will also hurt performance (due to the lack of caching effects).

**Comment by vhirtham** [comment]
@marscher Seems like the quality standards somehow carry over to other files now. Are you still on it?

File: weldx/config.py
Comment (vhirtham) [inline weldx/config.py]: I removed the changes because the loaded quality standard is now kept active for some reason. This causes validation errors in the follow-up tests. `wx_metadata` is now a required property in all other test files. So the behavior gets somehow worse. I think we should tackle this issue in another PR so that it doesn't prevent us from merging this PR.

**Review by CagtayFabry — APPROVED** [review]
🚀

File: tutorials/quality_standards.ipynb
Comment (vhirtham) [inline tutorials/quality_standards.ipynb:387]:
```suggestion
    "Its id is the manifest file's filepath relative to the `resource` directory with a preceding `http://`.\n",
```

**Pull request merged**
